File: src/main/java/gargoyle/ct/pref/CTPreferencesManager.java
package gargoyle.ct.pref;

import gargoyle.ct.pref.impl.prop.CTPrefProperty;
import org.jetbrains.annotations.NotNull;

import java.util.Set;

public interface CTPreferencesManager {
    void addPropertyChangeListener(CTPropertyChangeListener listener);

    @NotNull <T> CTPrefProperty<T> getProperty(String name);

    @NotNull <E extends Enum<E>> CTPrefProperty<E> getProperty(Class<E> type);

    @NotNull Set<String> getPropertyNames();

    boolean hasProperty(String name);

    void removePropertyChangeListener(CTPropertyChangeListener listener);
}

File: src/main/kotlin/gargoyle/ct/pref/CTPreferencesManager.kt
package gargoyle.ct.pref

import gargoyle.ct.pref.impl.prop.CTPrefProperty
import kotlin.reflect.KClass

interface CTPreferencesManager {
    fun addPropertyChangeListener(listener: CTPropertyChangeListener<*>)
    fun <T : Any> getProperty(name: String): CTPrefProperty<T>
    fun <E : Enum<E>> getProperty(type: KClass<E>): CTPrefProperty<E>
    val propertyNames: Set<String>
    fun hasProperty(name: String): Boolean
    fun removePropertyChangeListener(listener: CTPropertyChangeListener<*>)
}
